core/src/main/java/jenkins/install/SetupWizard.java

```diff
@@ -244,34 +245,33 @@ public HttpResponse doCreateAdminUser(StaplerRequest req, StaplerResponse rsp) t
                 admin.delete(); // assume the new user may well be 'admin'
             }
             
-            User u = securityRealm.createAccountByAdmin(req, rsp, "/jenkins/install/SetupWizard/setupWizardFirstUser.jelly", null);
-            if (u != null) {
-                if(admin != null) {
-                    admin = null;
-                }
-                
-                // Success! Delete the temporary password file:
-                try {
-                    getInitialAdminPasswordFile().delete();
-                } catch (InterruptedException e) {
-                    throw new IOException(e);
-                }
-                
-                InstallUtil.proceedToNextStateFrom(InstallState.CREATE_ADMIN_USER);
-                
-                // ... and then login
-                Authentication a = new UsernamePasswordAuthenticationToken(u.getId(),req.getParameter("password1"));
-                a = securityRealm.getSecurityComponents().manager.authenticate(a);
-                SecurityContextHolder.getContext().setAuthentication(a);
-                CrumbIssuer crumbIssuer = Jenkins.getInstance().getCrumbIssuer();
-                JSONObject data = new JSONObject();
-                if (crumbIssuer != null) {
-                    data.accumulate("crumbRequestField", crumbIssuer.getCrumbRequestField()).accumulate("crumb", crumbIssuer.getCrumb(req));
-                }
-                return HttpResponses.okJSON(data);
-            } else {
-                return HttpResponses.okJSON();
+            User u = securityRealm.createAccountFromSetupWizard(req);
+            if(admin != null) {
+                admin = null;
+            }
+
+            // Success! Delete the temporary password file:
+            try {
+                getInitialAdminPasswordFile().delete();
+            } catch (InterruptedException e) {
+                throw new IOException(e);
+            }
+
+            InstallUtil.proceedToNextStateFrom(InstallState.CREATE_ADMIN_USER);
+
+            // ... and then login
+            Authentication a = new UsernamePasswordAuthenticationToken(u.getId(),req.getParameter("password1"));
+            a = securityRealm.getSecurityComponents().manager.authenticate(a);
+            SecurityContextHolder.getContext().setAuthentication(a);
+            CrumbIssuer crumbIssuer = Jenkins.getInstance().getCrumbIssuer();
+            JSONObject data = new JSONObject();
+            if (crumbIssuer != null) {
+                data.accumulate("crumbRequestField", crumbIssuer.getCrumbRequestField()).accumulate("crumb", crumbIssuer.getCrumb(req));
             }
+            return HttpResponses.okJSON(data);
+        } catch (AccountCreationFailedException e) {
+            rsp.setStatus(400);
```

Comment:
Why `400`? According to what I see `AccountCreationFailedException` is for validating the input data, right?

Reply:
Yeah. What other status code would you suggest? According to this [Stackoverflow question](https://stackoverflow.com/questions/3050518/what-http-status-response-code-should-i-use-if-the-request-is-missing-a-required)'s answers (and [this](https://stackoverflow.com/a/14127995/1117552)), `400` may be a viable choice, but that technically indicates a syntax error. However, it seems that most APIs use it for any kind of input error, lacking a better option. `422` also goes in the direction of what we're trying to say, but that's not in the HTTP standard, WebDAV only.

Generally, I think `400` is definitely better than `200` here because the request was not "ok". Ideally, I'd be returning the error message as JSON, but sadly I can't do anything with that on the client since I can't modify the contents of the `iframe`.

`409` also seems to be a possible choice, but that doesn't describe our error better than `400`.

Reply:
@xxyy I agree. What do you think about `422`? [ref.](https://www.greenbytes.de/tech/webdav/rfc4918.html#rfc.section.11.2)

Reply:
@recena We can settle on this, it being not part of the initial HTTP/1.1 probably won't be a problem in practice.

Reply:
@xxyy To be honest, I'm not 100% sure about what the most used. What do you think about this description? https://httpstatuses.com/422

Reply:
@recena the description seems like it's more appropriate than the general 400. In my  experience, I've usually only seen 400, not 422. But honestly, it doesn't really matter - we know it's a client error so 4xx, and the frontend only cares about the fact that it's > 200. I've already changed it to 422 in case that wasn't clear.